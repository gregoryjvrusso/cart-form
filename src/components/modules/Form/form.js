import React, { useState } from "react";
import { arrayOf, func, shape, string } from "prop-types";
import Button from "../../elements/Button";
import Flex from "../../elements/Flex";
import Input from "../../elements/Input";
import InputSelected from "../../elements/InputSelected";
import regex from "../../../constants/regex-form";
import {
  onInputChange,
  onInputBlur,
  onInputSelectBlur,
  handleSubmit,
} from "../../../utils/formEvent";
import { Container, FlexForm } from "./form.styles";

const Form = ({ dataSelect, ...props }) => {
  const errorData = {
    card: false,
    name: false,
    date: false,
    cvv: false,
    subdivision: false,
  };
  const [error, setError] = useState(errorData);

  return (
    <Flex width={"100%"}>
      <Container onSubmit={(e) => handleSubmit(e, error)} width={"100%"} id="form">
        <FlexForm>
          <Input
            name="card"
            textLabel="Número do cartão"
            value={props.card}
            setValue={props.setCard}
            error={error}
            setError={setError}
            messageError="Número de cartão inválido"
            onInputChange={onInputChange}
            onInputBlur={onInputBlur}
            mask="9999 9999 9999 9999"
            regex={regex}
          />
        </FlexForm>
        <FlexForm>
          <Input
            name="name"
            textLabel="Nome"
            value={props.name}
            setValue={props.setName}
            error={error}
            setError={setError}
            messageError="Insira seu nome completo"
            onInputBlur={onInputBlur}
            onInputChange={onInputChange}
            regex={regex}
          />
        </FlexForm>
        <FlexForm>
          <FlexForm>
            <Input
              name="date"
              textLabel="Validade"
              value={props.date}
              setValue={props.setDate}
              error={error}
              setError={setError}
              messageError="Data inválida"
              onInputBlur={onInputBlur}
              onInputChange={onInputChange}
              mask="99/99"
              regex={regex}
            />
          </FlexForm>
          <FlexForm ml={2}>
            <Input
              name="cvv"
              textLabel="CVV"
              value={props.cvv}
              setValue={props.setCvv}
              error={error}
              setError={setError}
              messageError="Código inválido"
              onInputBlur={onInputBlur}
              onInputChange={onInputChange}
              mask="999"
              regex={regex}
            />
          </FlexForm>
        </FlexForm>
        <FlexForm>
          <InputSelected
            onInputBlur={onInputSelectBlur}
            onInputChange={onInputChange}
            value={props.subdivision}
            setValue={props.setSubdivision}
            name="subdivision"
            options={dataSelect}
            textLabel="Quantidade de parcelas"
            error={error}
            setError={setError}
            messageError="Insira o número de parcelas"
          />
        </FlexForm>
        <Flex justifyContent={"flex-end"} mt={3}>
          <Flex width={["100%", "100%", "50%"]}>
            <Button type="submit" form="form" value="Submit">
              CONTINUAR
            </Button>
          </Flex>
        </Flex>
      </Container>
    </Flex>
  );
};

Form.propTypes = {
  dataSelect: arrayOf(
    shape({
      value: string,
      text: string,
    })
  ).isRequired,
};

Form.displayName = "Form";

export default Form;
