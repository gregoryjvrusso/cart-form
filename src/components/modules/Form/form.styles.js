import styled from "styled-components";
import Flex from "../../elements/Flex";

export const FlexForm = styled(Flex)`
  position: relative;
  width: 100%;
`;

export const Container = styled.form`
  width: 100%;
`;
