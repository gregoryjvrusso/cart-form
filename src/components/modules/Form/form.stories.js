import React from "react";
import { storiesOf } from "@storybook/react";
import { ThemeProvider } from "styled-components";
import theme from "../../../common/theme";

import Form from "../Form";
const mockData = [
  {
    value: "",
    text: "",
  },
  {
    value: "1",
    text: "Parcelar em 1 vez",
  },
  {
    value: "2",
    text: "Parcelar em 2 vezes",
  },
  {
    value: "3",
    text: "Parcelar em 3 vezes",
  },
  {
    value: "4",
    text: "Parcelar em 4 vezes",
  },
  {
    value: "5",
    text: "Parcelar em 5 vezes",
  },
  {
    value: "6",
    text: "Parcelar em 6 vezes",
  },
];

storiesOf("Modules/Form", module).add("default", () => {
  return (
    <ThemeProvider theme={theme}>
      <Form dataSelect={mockData} />
    </ThemeProvider>
  );
});
