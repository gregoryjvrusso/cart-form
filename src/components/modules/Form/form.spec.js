import React, {useState} from "react";
import { render, fireEvent } from "@testing-library/react";
import Form from "../Form";

const mockData = [
  {
    value: "",
    text: "",
  },
  {
    value: "1",
    text: "Parcelar em 1 vez",
  },
  {
    value: "2",
    text: "Parcelar em 2 vezes",
  },
  {
    value: "3",
    text: "Parcelar em 3 vezes",
  },
  {
    value: "4",
    text: "Parcelar em 4 vezes",
  },
  {
    value: "5",
    text: "Parcelar em 5 vezes",
  },
  {
    value: "6",
    text: "Parcelar em 6 vezes",
  },
];

describe("Form component", () => {
  it("deve conseguir renderizar [snapshot]", async () => {
    const { container } = render(<Form dataSelect={mockData} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("deve conseguir renderizar [snapshot]", async () => {
    const { getAllByTestId } = render(<Form dataSelect={mockData} />);
    const inputs = getAllByTestId('input');
    expect(inputs[0].value).toBe('');
    fireEvent.change(inputs[0], { target: { value: '0032 2323 3232 1212' } });
    expect(inputs[0].value).toBe('0032 2323 3232 1212');
  });
});
