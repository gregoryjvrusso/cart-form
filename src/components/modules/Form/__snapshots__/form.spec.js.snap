// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component deve conseguir renderizar [snapshot] 1`] = `
<div
  class="sc-AxirZ fBnoWH sc-AxiKw bKsCdD"
  data-testid="flex"
  width="100%"
>
  <form
    class="sc-fzplWN jHAIBM"
    id="form"
    width="100%"
  >
    <div
      class="sc-AxirZ lgJgCc sc-AxiKw WlbxU sc-fzpans cKVhrP"
      data-testid="flex"
    >
      <div
        class="sc-AxirZ vJexS"
        data-testid="box"
        width="100%"
      >
        <input
          class="sc-AxhUy bFivQe"
          data-testid="input"
          id="card"
          name="card"
          placeholder="Número do cartão"
          regex="[object Object]"
          required=""
          type="input"
          value=""
        />
        <label
          class="sc-AxgMl eMytDd"
          for="card"
        >
          Número do cartão
        </label>
      </div>
    </div>
    <div
      class="sc-AxirZ lgJgCc sc-AxiKw WlbxU sc-fzpans cKVhrP"
      data-testid="flex"
    >
      <div
        class="sc-AxirZ vJexS"
        data-testid="box"
        width="100%"
      >
        <input
          class="sc-AxhUy bFivQe"
          data-testid="input"
          id="name"
          name="name"
          placeholder="Nome"
          regex="[object Object]"
          required=""
          type="input"
          value=""
        />
        <label
          class="sc-AxgMl eMytDd"
          for="name"
        >
          Nome
        </label>
      </div>
    </div>
    <div
      class="sc-AxirZ lgJgCc sc-AxiKw WlbxU sc-fzpans cKVhrP"
      data-testid="flex"
    >
      <div
        class="sc-AxirZ lgJgCc sc-AxiKw WlbxU sc-fzpans cKVhrP"
        data-testid="flex"
      >
        <div
          class="sc-AxirZ vJexS"
          data-testid="box"
          width="100%"
        >
          <input
            class="sc-AxhUy bFivQe"
            data-testid="input"
            id="date"
            name="date"
            placeholder="Validade"
            regex="[object Object]"
            required=""
            type="input"
            value=""
          />
          <label
            class="sc-AxgMl eMytDd"
            for="date"
          >
            Validade
          </label>
        </div>
      </div>
      <div
        class="sc-AxirZ iJPJzN sc-AxiKw WlbxU sc-fzpans cKVhrP"
        data-testid="flex"
      >
        <div
          class="sc-AxirZ vJexS"
          data-testid="box"
          width="100%"
        >
          <input
            class="sc-AxhUy bFivQe"
            data-testid="input"
            id="cvv"
            name="cvv"
            placeholder="CVV"
            regex="[object Object]"
            required=""
            type="input"
            value=""
          />
          <label
            class="sc-AxgMl eMytDd"
            for="cvv"
          >
            CVV
          </label>
        </div>
      </div>
    </div>
    <div
      class="sc-AxirZ lgJgCc sc-AxiKw WlbxU sc-fzpans cKVhrP"
      data-testid="flex"
    >
      <div
        class="sc-AxirZ vJexS"
        data-testid="box"
        width="100%"
      >
        <select
          class="sc-Axmtr egUstk"
          data-testid="input-select"
          name="subdivision"
          placeholder="Quantidade de parcelas"
          required=""
        >
          <option
            class="sc-AxmLO btQQyJ"
            data-testid="options"
            value=""
          />
          <option
            class="sc-AxmLO btQQyJ"
            data-testid="options"
            value="1"
          >
            Parcelar em 1 vez
          </option>
          <option
            class="sc-AxmLO btQQyJ"
            data-testid="options"
            value="2"
          >
            Parcelar em 2 vezes
          </option>
          <option
            class="sc-AxmLO btQQyJ"
            data-testid="options"
            value="3"
          >
            Parcelar em 3 vezes
          </option>
          <option
            class="sc-AxmLO btQQyJ"
            data-testid="options"
            value="4"
          >
            Parcelar em 4 vezes
          </option>
          <option
            class="sc-AxmLO btQQyJ"
            data-testid="options"
            value="5"
          >
            Parcelar em 5 vezes
          </option>
          <option
            class="sc-AxmLO btQQyJ"
            data-testid="options"
            value="6"
          >
            Parcelar em 6 vezes
          </option>
        </select>
        <label
          class="sc-fzozJi iivSCe"
        >
          Quantidade de parcelas
        </label>
      </div>
    </div>
    <div
      class="sc-AxirZ hPIDTq sc-AxiKw fveomK"
      data-testid="flex"
    >
      <div
        class="sc-AxirZ jdCCXI sc-AxiKw ewUtxs"
        data-testid="flex"
        width="100%,100%,50%"
      >
        <button
          class="sc-AxjAm eEzdfU"
          data-testid="button"
        >
          CONTINUAR
        </button>
      </div>
    </div>
  </form>
</div>
`;
